src/library/vm/interaction_state.h

```
@@ -0,0 +1,242 @@
+        }
+    }
+
+    static bool is_silent_exception(vm_obj const & ex) {
+        return is_constructor(ex) && cidx(ex) == 1 && is_none(cfield(ex, 0));
+    }
+
+    static vm_obj mk_result(vm_obj const & a, vm_obj const & s) {
+        lean_assert(is_State(s));
+        return mk_vm_constructor(0, a, s);
+    }
+
+    static bool is_result_exception(vm_obj const & r) {
+        return is_constructor(r) && cidx(r) == 1;
+    }
+
+    static bool is_result_success(vm_obj const & r) {
+        return is_constructor(r) && cidx(r) == 0;
+    }
+
+    static vm_obj get_result_value(vm_obj const & r) {
+        lean_assert(is_result_success(r));
+        return cfield(r, 0);
+    }
+
+    static vm_obj get_result_state(vm_obj const & r) {
+        lean_assert(is_result_success(r));
+        return cfield(r, 1);
+    }
+
+    static vm_obj mk_success(vm_obj const & a, State const & s) {
+        return mk_vm_constructor(0, a, to_obj(s));
+    }
+
+    static vm_obj mk_success(State const & s) {
+        return mk_success(mk_vm_unit(), s);
+    }
+
+    static vm_obj mk_exception(vm_obj const & fn, State const & s) {
+        return mk_vm_constructor(1, mk_vm_some(fn), mk_vm_none(), to_obj(s));
+    }
+
+    static vm_obj mk_silent_exception(State const & s) {
+        return mk_vm_constructor(1, mk_vm_none(), mk_vm_none(), to_obj(s));
+    }
+
+    static vm_obj mk_exception(vm_obj const & fn, vm_obj const & ref, State const & s) {
+        return mk_vm_constructor(1, mk_vm_some(fn), ref, to_obj(s));
+    }
+
+    static vm_obj mk_exception(throwable const & ex, State const & s) {
+        vm_obj _ex = lean::to_obj(ex);
+        vm_obj fn = mk_vm_closure(get_throwable_to_format_fun_idx(), 1, &_ex);
+        optional<expr> ref;
+        if (auto kex = dynamic_cast<ext_exception const *>(&ex))
+            ref = kex->get_main_expr();
+        else if (auto fex = dynamic_cast<formatted_exception const *>(&ex))
+            ref = fex->get_main_expr();
```

@leodemoura This code should now also handle `parser_error`s in order to propagate error positions. But I don't see how to do that since those exceptions contain only a position, not an `expr`. We could further abstract over the type of information stored in `interaction_monad.result` (`pos` for parser, `expr` for tactic), but I don't think we actually do anything with the `expr` right now apart from querying its position. Should we switch to just
```
| exception {} : option (unit → format) → option pos /- currently expr -/ → state → interaction_monad.result
```
?

---

Yes, switching to `pos` is good.

---

Hm. Does that mean we should switch `formatted_exception` and all its derived classes, too?

---

I now remember why we store `expr`.
The kernel has to report errors, but it doesn't know anything about position information.
Before Lean 3, only the modules at `front_end/lean` were capable of retrieving position information.
We now have `scope_pos_info_manager`, and the thread local `get_pos_info_provider`.
However, it is defined in library.
I'm not sure what is the right solution.

We could start using `pos_info` in all these exceptions, but we would have to move `scope_pos_info_manager` to the kernel, and also modify `ext_exception` defined in the kernel.

---

I think the simplest solution is to move `scope_pos_info_provider.*` to the kernel, and use `pos` instead of `expr` in all these exceptions.

---

Okay, I'll do the change tomorrow. That should fix the last failing test.

---

@leodemoura What is the difference between `parser_error` and `parser_exception`?

---

The parser has two modes when an error is detected
 1- report error & continue
 2- throw an exception (and interrupt its execution)

Mode 2 is not used very often. It is useful for exposing the parser API.

We use `parser_error` to report errors like `invalid expression, '(' expected`. These exceptions are handled by the parser itself. The `parser_exception` is used to communicate errors to the module that invoked the parser.

---

I see, though I'm not sure I see the difference very clearly in the code right now. I would expect the scanner to throw `parser_error`, and both exceptions seem to be swallowed the same by `protected_call`.

Either way, I've adjusted all the exceptions now. `protected_call` will now catch and report all `exception_with_pos`, such as `parser_error`, or `formatted_exception` from running a Lean parameter parser.